### DISCUSSION TITLE
UI issues

### DISCUSSION THREAD

**Discussion by kybos** [muse-sequencer/muse#1164] [Team Posts]
As announced in the other thread below, a quick collection of Ui related issues that I think should be changed/improved. Feel free to comment and contradict ;-).

1. The only deep and conceptual problem I see is the track colour handling (part specific, "current colour" concept). I suppose it stems from the original MusE, IMO one of the few things that make absolutely no sense, the colour should be track specific.

2. Mixer strip: Right click menu on track name:
Almost all menu points are the same as in the global settings, so as a user I would expect that here they are strip-specific, but they are not, they are really just a duplicates of the global settings. Not nice, but not critical, as they work and show immediate effect.
Except: Visible audio effects. No global setting exists, and changing it does seemingly nothing, so it looks like a bug to the user. It should at least go to the global settings (only), with a note/tooltip that it only takes effect for newly opened mixer strips.

3. The in/out icons are nice, but they break the (horizontal/vertical) symmetry without a frame/grid or something similar.

4. The "palettes" in midi strip: A and B are no sensible texts (and, still worse, "Palette A/B" no sensible tooltips). Palette A contains two unrelated areas, midi instrument definition and midi parameters like chorus etc. (palette B is more consistent). So logically, there should perhaps be even 3 "tabs".

5. Pan slider is wrongly used, as pan function is symmetrical based on the midpoint.

6. All the icons at the bottom of the mixer strip: They are positioned quite randomly and jump depending on the track type. Again, missing grid/symmetry.

7. All the icons again + mono/stereo + pre-fader listening: The light-blue colour strongly suggests that they are inactive, but not: Most of them are, but Mute and Prefader are active. This is completely illogical and confusing.

8. The "empty" button on the bottom of the midi strips looks like a bug, there can be nothing like a button without a text/icon.

9. My favourite ;-) : The "track info" button/icon: It should not be there at all, it's inconsistent with other views and (despite my efforts) still looks ugly and misplaced. Perhaps could be replaced by a key shortcut at least...

To be continued ;-) ...

**Comment by terminator356** [comment]
I try to avoid long-winded replies these days because everything gets lost in noise. But here goes...

1: I tried so many times to find a solution that keeps what we have so far
 while being track-specific at part creation time. Not possible I think.
Let's see if we can solve this once and for all, maybe right now with this proposal:
I propose we give each track a 'current colour' instead of having a global 'current colour'.
At first, by default, each track creates new parts in the **track's colour**.
But if the user changes the track's current colour, new parts **on that track** are created in that colour.
What about new tracks? Possibly, we could carry over a track type's current colour
 to any new tracks **of that type** so that new parts on that track are the same as the last colour used
 on another track of that type.
Otherwise, any new tracks will have to start off by creating new parts in the track's colour.
Lastly, the colour popup menu should highlight the current track colour.
Currently it does not even highlight the current global colour.
I have long planned to put a drop-down colour box on the toolbars for quick access.
I had ideas to have a colour selector square there as well so that **any** colour could be chosen,
 but that does not jive with highlighting the current colour in the colour popup menu.
Actually come to think of it, neither does track colour, which is not in the list of colours.
We might have to change some of those pre-defined colours, or **append** some new ones
 to match track colours. A further wrinkle is that track colours are user definable (in Appearance).

2: They are exactly the same as global settings on purpose. Convenience of at-hand settings.
I don't want to go to global settings just to change something on a strip. And then click OK.
I made it that way for the immediacy of their effect. Why is this so bothersome?
I _might_ counter that more things from global settings should be available at-hand all over the place.
Visible audio effects "does seemingly nothing"? It immediately changes the number of effects shown.
Please explain.

3: Symmetry? Please explain. As in 6 below, I avoided grids and used hbox because I would rather
 the icons squeeze together naturally instead of leaving empty grid space.

** Please note: in 3 and 6 if you are referring to the symmetry in an **expanded** strip, I
 explained some of that recently but I'll clarify more:
The buttons do not move with the strip's expansion splitter.
I carefully considered that. I felt it was better to leave them in place rather than have
 many strips of different width each looking different than the others.
This way users always know where the buttons are consistently going to be, and they're close together.

4: It took many years to get to this point. It is always a work in progress.
A look back through the years at previous designs shows why we're here.
We didn't have a way to stuff many things vertically. Now we do (the effects rack with a scrollbar).
So I had to put the stuff in two tabs. I had no choice at the time.
I am not proud of how it looks now. It's a low point. But it will change.
The choice of two tabs made sense at the time simply because it made everything line up
 vertically. If the three variation, reverb and chorus controls were moved to another tab,
 then the midi strip meters were going to be weirdly much longer than the audio strips.
So I tried to at least make everything line up as much as possible.
Again, I mentioned that I am working towards allowing users to put whatever controls
 they want in there. The mechanism is there. We'll see how the strip turns out and
 what goes where when it's done.

5: Please explain. A visible dividing line in the middle? Not as easy as it seems.
The compact slider control classes, which are designed for general use anywhere,
 have no knowledge of 'pan' vs. other controllers. We **could** arbitrarily draw a line
 at position '0'. Fine enough, but then all such controls would draw that line at '0',
 and not every controller is symmetric. We would end up with some controls where the line 
 is not in the middle or worse, all the way to one side.
I suppose it could be cured by adding yet another feature to the compact slider classes
 where the line is optional and the creator decides whether it should be shown.
One more thing: Drawing a line on top of the slider means the loss of one visible
 position - at '0', how do we know visibly if it's really on '0', or '-1' or '1'?
I suppose the line could be slightly transparent to allow the slider to show its edge at position '0'.

6: They are not positioned randomly. Each icon has a definite place where it goes
 **if** it is visible for a strip type. There may be some slight horizontal shifting going on
 since I did purposely use an hbox **not** a grid. A grid would force a big ol' ugly empty space
 where an icon would have been placed if it were not hidden. So an hbox places them
 more naturally squeezed together.

7: Again, the idea with that, is that in a fully populated mixer, I only wanted things that are active to
 stand out and alert us that they are active. It's harder to tell that from an icon that simply changes shape.
Mute and pre-fader are 'active' when they are 'on'. Hence the red line through mute when 'active'
 and the pre-fader 'comes alive' when active.
That's the operative phrase: 'Comes alive'. The blue icons are designed to be unobtrusive and
 not distracting and only come live when something is 'active'.
I concede that mute and pre-fader were the most troublesome and had me questioning
 the logic and whether my entire 'come alive' scheme would work.
But the logic is correct ! The icons might be better, but it's tricky and that's what I came up with.

8: It's a placeholder. If it is removed the entire strip will shift down. And empty space might look worse.
The intention has always been to put midi automation there, just like the audio. No one's done it yet.

9: Well, I'll leave that one for another day. It's been there forever. I just tried to keep what we had.
The alternative plan was to try and add it to all the editors.
I've got my eye on that but there was no time to do it before release, just fix the drum editor.

**Comment by kybos** [comment]
2.
> It immediately changes the number of effects shown.

Not here - a theme-specific problem or something?

5. I know it's all not easy, there is no criticism intended. This whole collection of mine is just trying to see things from a user perspective (because in fact I am a relatively new user...). For the Pan I would possibly even prefer if it always showed as a knob, the on-sided slider not being suitable. Maybe not so bad, as it is the only control down there...

6. Optimally the icons should have their fixed place independently of track type. Again I understand why you did it from a dev perspective, but as a user it irritates me ;-). Ugly empty spaces you have anyway, just look at Group/Input/Aux tracks. Just an idea: To avoid adding another row when there are 5 icons, the Track off icon could go to the top, between the In/Out icons. Not optimal either, but at least consistent for all track types. Or no icon for it at all, put it into the context menu.

7. This one I don't understand even as a dev ;-), but maybe I am missing something. I thought the pre-fader is active as a default, but it has the same blue colour that in the other icons signals inactivity.

**Comment by terminator356** [comment]
2: Hm, you should immediately see the number of visible slots change, on all the strips,
 even the left-side ones in the editors. Wow, that's not happening there? Ugh... Sorry for confusion.
Each of the racks is supposed to respond and immediately resize.
What can you tell me about your setup? Are you using a stylesheet? What selected Qt theme?
What happens when everything is reset to defaults?

5: I see. The one-sided pan slider not tickling your fancy? The compact slider class has a few tricks
 up its sleeve, things which aren't even shown (such as an actual thumb). I'll see if I can tweak them.
Kick me later or enter a bug report so I don't forget.

7: Yep, pre-fader is inactive (off) by default. That's why it's blue. Pre-fader means that
 the audio is taken from _before_ the volume slider, ie. the volume slider will have no effect at all.
The icon itself, meh... could not think of what else to put there except a picture of a slider ;-)

**Comment by kybos** [comment]
2. I use Fusion as Qt theme (it's the only one that really works on a GTK desktop). It happens even on the default MusE theme, without stylesheet. I will test later, also with KDE, when I am on my main PC. 

7. Pre-fader: Yes, of course the function is inactive as a default, see how I am confused myself ;-). It's inactive, then you click on it, and it's double inactive? Because this is what the same icon, but stricken-through, means for a user - the function has just been deactivated. 
I think I understand what you were trying to achieve after all your explanations, but you can't expect a normal user to catch that.
Remedy: Include the strike-through line also in the inactive icon (in blue of course), and colour the active icon completely, also the blue parts. Then it would be clear. Or: Change the icon, perhaps a small horizontal arrow to the left of the fader to symbolize pre-fader.
The problem with the Mute icon is very similar.

**Comment by terminator356** [comment]
Oops. My mistake. I've caused more confusion.
I totally forgot that the pre-fader icon has a line through it when active !
I thought it 'came alive' with a full graphic there.
Yeah, now I remember why it was so hard to make an icon there.

**Comment by kybos** [comment]
2. Tested with KDE, there it works indeed. But not with my system (Mint/Cinnamon), even if I delete the settings, strange...

**Comment by terminator356** [comment]
OK Thanks. That's worrisome.
Makes me wonder how many other things that should just work, don't for you.
Maybe you're seeing weird things that I had no idea weren't working properly.
I lack space on my drives for another distro.

If you use say KDevelop, I'll try to find a few breakpoints you can check, I looked
 briefly the other day but got side-tracked. Today I may be busy taking care of a few things...

**Comment by kybos** [comment]
I looked into it, my impression is that the problem is with Qt 5.9 that I have here. The size hint for the rack seems not to be updated properly, or too late (I tried all the usual updates/invalidates/adjustSizes etc. in the configChanged method of the audio strip, no way). I've been struggling with a very similar problem for weeks, in another context, so I'm not so surprised. And the setup of the mixer strip is quite a complex thing... 
So I guess we leave it at that. It's not critical, not a function used very often, and it will most probably disappear with later Qt versions.

**Comment by spamatica** [comment]
Hi guys, good discussion. 
I don't have that much to contribute to it currently. Most of layout issues I don't think I have noticed, might be it is style specific... which is bothersome... or that I simply don't care enough ;)

Regarding 9. Track info. why not just remove it and add a checkable menu item in, say, the View menu? Everything doesn't have to have a button.

Also, regarding the Transport panel (which is currently ontop of this window where I'm currently writing). 
An easy fix (in waiting for moving the functionality to better places) is to simply make the always-on-top property optional.

**Comment by kybos** [comment]
> Most of layout issues I don't think I have noticed, might be it is style specific... which is bothersome... or that I simply don't care enough ;)

I do, and I know many other people do too ;-).

> Regarding 9. Track info. why not just remove it and add a checkable menu item in, say, the View menu? Everything doesn't have to have a button.

I fully agree. I think it should go to the Display menu, because it's view-specific (available only for the Arranger). I personally would remove it completely, I see no need for it (the strip can be hidden with the splitter anyway, and it's missing in all the other views except Arranger), but Tim does not like that ;-).

> Also, regarding the Transport panel (which is currently ontop of this window where I'm currently writing).
> An easy fix (in waiting for moving the functionality to better places) is to simply make the always-on-top property optional.

A good idea.

**Comment by spamatica** [comment]
Hello.

sorry, been occupied by so many other things... and still am, just wanted
to show a sign of life.

Den sön 29 mars 2020 kl 11:20 skrev kybos <notifications@github.com>:

> Most of layout issues I don't think I have noticed, might be it is style
> specific... which is bothersome... or that I simply don't care enough ;)
>
> I do, and I know many other people do too ;-).
>

It was mostly tongue in cheek, I do care! :)
But some of these things has gone past me so either I'm not affected or my
eyesight is not good enough.


> Regarding 9. Track info. why not just remove it and add a checkable menu
> item in, say, the View menu? Everything doesn't have to have a button.
>
> I fully agree. I think it should go to the Display menu, because it's
> view-specific (available only for the Arranger). I personally would remove
> it completely, I see no need for it (the strip can be hidden with the
> splitter anyway, and it's missing in all the other views except Arranger),
> but Tim does not like that ;-).
>
I guess it could be confusing if there is no evident way to add it back if
it was collapsed by mistake.
Another solution would be for the collapsed strip to have a "handle" left
when collapsed.

Oh, I see now that it is separate. The hide/show button does not bring the
collapsed strip back. I think we should settle on one or the other way of
doing it. Having both can be confusing.

Also, regarding the Transport panel (which is currently ontop of this

**Comment by terminator356** [comment]
> I guess it could be confusing if there is no evident way to add it back if it was collapsed by mistake.

Yup. And as I said, when it is expanded 'out there', not having a button to collapse it,
 then re-expand it to its original position means users must fiddle with the handle 
 to get it to go back to where they left it.

> The hide/show button does not bring the collapsed strip back.

Oops. ++1 For finding a way to bring it back.

**Comment by kybos** [comment]
So obviously there is need for both variants.

1. The standard way, using the splitter, which every user knows and understands. Splitter handle is always visible, even if the mixer strip is moved to the left and hidden that way. It can be restored again using the handle.

2. A special toggle function to close the strip completely/restore it exactly to its previous width. I am sure most users won't even notice the pseudo-button down there (or the former TrackInfo button for that matter), so I would consider that an expert function. As such, it makes more sense to put it in a menu and assign a key shortcut (e.g. F3) to it, which can be used much more efficiently than pointing the mouse to the small button every time.
